### PR TITLE
Enable specifying output assembly name in compiler options

### PR DESCRIPTION
**Pull request by moozzyk**
Addresses #1797

### CODE REVIEW

**Comment by moozzyk** [comment]
Recreating https://github.com/dotnet/cli/pull/1807 to use new CI

File: TestAssets/TestProjects/LibraryWithOuputAssemblyName/project.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (brthor) [inline TestAssets/TestProjects/LibraryWithOuputAssemblyName/project.json]: nit: typo in the directory here

Comment (moozzyk) [inline TestAssets/TestProjects/LibraryWithOuputAssemblyName/project.json]: will fix

File: test/dotnet-compile.Tests/CompilerTests.cs [hunk 2]
```diff
@@ -167,6 +167,24 @@ public void DepsDependencyContextIsValidOnBuild()
...
+                    .WithLockFiles();
+
+            var root = testInstance.TestRoot;
+            var outputDir = Path.Combine(root, "bin");
+            var testProject = ProjectUtils.GetProjectJson(root, "LibraryWithOutputAssemblyName");
+            var buildCommand = new BuildCommand(testProject, output: outputDir, framework: DefaultFramework);
+            var result = buildCommand.ExecuteWithCapturedOutput();
+            result.Should().Pass();
+
+            new DirectoryInfo(outputDir).Should().HaveFiles(new [] { "MyLibrary.dll" });
```
Comment (pakrym) [inline test/dotnet-compile.Tests/CompilerTests.cs]: Can you check that misc files like *.deps also have correct names for executable?

Comment (moozzyk) [inline test/dotnet-compile.Tests/CompilerTests.cs]: OK - I assume that I will have to use my package as a dependency to check that. Correct? Any other files/scenarios you would recommend verifying?

Comment (pakrym) [inline test/dotnet-compile.Tests/CompilerTests.cs]: Just make it an app not a library

File: src/Microsoft.DotNet.Compiler.Common/CommonCompilerOptionsExtensions.cs [hunk 3]
```diff
@@ -205,6 +212,11 @@ public static IEnumerable<string> SerializeToArgs(this CommonCompilerOptions opt
                 args.Add(s_generateXmlDocumentation.ToLongArg(generateXmlDocumentation));
             }
 
+            if (outputName != null)
```
Comment (troydai) [inline src/Microsoft.DotNet.Compiler.Common/CommonCompilerOptionsExtensions.cs]: Could it be empty?

**Comment by troydai** [comment]
:shipit:

File: test/dotnet-compile.Tests/CompilerTests.cs (hunk 2 above)
Comment (moozzyk) [inline test/dotnet-compile.Tests/CompilerTests.cs]: @pakrym I tried it and I can see that deps now has: 
```
        "runtime": {
          "MyApp.dll": {}
        }
```
Is this what needs to be checked, right?

(on a separate note when creating an app the dll is named as per the requested output name but the .exe itself has the package name - I think we want .exe to have the same name as specified in the outputName right @davidfowl ? if so, I will investigate this separately...)

Comment (moozzyk) [inline test/dotnet-compile.Tests/CompilerTests.cs]: Actually I also see this in the deps file: 
  "libraries": {
    "LibraryWithOutputAssemblyName/1.0.0": {
      "type": "project",
      "serviceable": false,
      "sha512": ""
    },

and also the same dll on the disk so am not sure if this is right...

Comment (pakrym) [inline test/dotnet-compile.Tests/CompilerTests.cs]: > (on a separate note when creating an app the dll is named as per the requested output name but the .exe itself has the package name - I think we want .exe to have the same name as specified in the outputName right @davidfowl ? if so, I will investigate this separately...)

That's what I meant, .exe, .deps.json should be named as your assembly name.

Comment (moozzyk) [inline test/dotnet-compile.Tests/CompilerTests.cs]: OK. I will look into it but I would prefer to do this separately. Tools are not executable so merging this change would unblock further work on tools while I am working on executables.

File: src/Microsoft.DotNet.Compiler.Common/CommonCompilerOptionsExtensions.cs (hunk 3 above)
Comment (moozzyk) [inline src/Microsoft.DotNet.Compiler.Common/CommonCompilerOptionsExtensions.cs]: Empty values are being checked elsewhere. The error message is not great as it does relate exactly to the name from project.json (`option --output-name requires a value`) - but this is consistent with other options (I think I checked the xmlDoc option)

File: test/dotnet-compile.Tests/CompilerTests.cs (hunk 2 above)
Comment (davidfowl) [inline test/dotnet-compile.Tests/CompilerTests.cs]: We should really fix that though, I rather not have anything like this be broken...

Comment (moozzyk) [inline test/dotnet-compile.Tests/CompilerTests.cs]: In that case I will update the PR with another commit that fixes things for exes

**Comment by moozzyk** [comment]
:up::date:
- added support for outputName in apps
- updated publish to write the correct .exe name

**Comment by pakrym** [comment]
Can we abstract `compilationOptions.OutputName ?? context.ProjectFile.Name` somewhere? It has 4 usages in this PR

**Comment by moozzyk** [comment]
I thought about this but am not sure if it is right to mix compilation options with ProjectContext...

**Comment by pakrym** [comment]
Maybe we can set it to CompilationOptions

**Comment by pakrym** [comment]
Accidental close, sorry

**Comment by moozzyk** [comment]
Folded `ProjectFile.Name ?? compilerOptions.OutputName` to `compilerOptions.OutputName`.

File: src/Microsoft.DotNet.ProjectModel/RuntimeOutputFiles.cs [hunk 4]
```diff
@@ -37,8 +37,12 @@ public string Executable
                     // The executable is a DLL in this case
                     extension = FileNameSuffixes.DotNet.DynamicLib;
                 }
+                
+                var compilationOptions = Project.GetCompilerOptions(Framework, Configuration);
 
-                return Path.Combine(BasePath, Project.Name + extension);
+                return Path.Combine(
+                    BasePath,
+                    (compilationOptions.OutputName ?? Project.Name) + extension);
```
Comment (pakrym) [inline src/Microsoft.DotNet.ProjectModel/RuntimeOutputFiles.cs]: What about this usage?

Comment (moozzyk) [inline src/Microsoft.DotNet.ProjectModel/RuntimeOutputFiles.cs]: Fixing. Must have had different pattern and VS Code did not find it...

**Comment by troydai** [comment]
Still :shipit:

**Comment by moozzyk** [comment]
Some CI failures are not related to this change

CentOs:
the tool /mnt/resource/j/workspace/dotnet_cli/rel_1.0.0/debug_centos7.1_x64_prtest/.dotnet_stage0/Linux/share/dotnet/cli/bin/dotnet cannot be found

Windows:
20:41:35 dotnet : The term 'dotnet' is not recognized as the name of a cmdlet, 
20:41:35 function, script file, or operable program. Check the spelling of the name, or 
20:41:35 if a path was included, verify that the path is correct and try again.

**Comment by moozzyk** [comment]
Rebased on the latest changes but Ubuntu CI still fails: 

10:29:51 Restoring Build Script projects...
10:29:51 /opt/code/scripts/run-build.sh: line 105: dotnet: command not found

**Comment by pakrym** [comment]
:shipit:

**Pull request merged**